src/contrasts.jl

```diff
@@ -0,0 +1,215 @@
+# will behave identically in creating modelmatrix columns
+function Base.:(==)(
+    a::ContrastsMatrix{C,T}, b::ContrastsMatrix{C,T}
+) where {C<:AbstractContrasts,T}
+    return a.matrix == b.matrix && a.termnames == b.termnames && a.levels == b.levels
+end
+
+function Base.hash(a::ContrastsMatrix{C}, h::UInt) where {C}
+    return hash(C, hash(a.matrix, hash(a.termnames, hash(a.levels, h))))
+end
+
+"""
+An instantiation of a contrast coding system for particular levels
+
+This type is used internally for generating model matrices based on categorical
+data, and **most users will not need to deal with it directly**.  Conceptually,
+a `ContrastsMatrix` object stands for an instantiation of a contrast coding
+*system* for a particular set of categorical *data levels*.
+
+If levels are specified in the `AbstractContrasts`, those will be used, and likewise
+for the base level (which defaults to the first level).
+
+# Constructors
+
+```julia
+ContrastsMatrix(contrasts::AbstractContrasts, levels::AbstractVector)
+ContrastsMatrix(contrasts_matrix::ContrastsMatrix, levels::AbstractVector)
+```
+
+# Arguments
+
+* `contrasts::AbstractContrasts`: The contrast coding system to use.
+   We only use `DummyCoding()`
+
+* `levels::AbstractVector`: The levels to generate contrasts for.
+
+* `contrasts_matrix::ContrastsMatrix`: Constructing a `ContrastsMatrix` from
+  another will check that the levels match.  This is used, for example, in
+  constructing a model matrix from a `ModelFrame` using different data.
+"""
+function ContrastsMatrix(
+    contrasts::C, levels::AbstractVector{T}
+) where {C<:AbstractContrasts,T}
+
+    # if levels are defined on contrasts, use those, validating that they line up.
+    # what does that mean? either:
+    #
+    # 1. DataAPI.levels(contrasts) == levels (best case)
+    # 2. data levels missing from contrast: would generate empty/undefined rows.
+    #    better to filter data frame first
+    # 3. contrast levels missing from data: would have empty columns, generate a
+    #    rank-deficient model matrix.
+    c_levels = something(DataAPI.levels(contrasts), levels)
+    if eltype(c_levels) != eltype(levels)
+        throw(
+            ArgumentError(
+                "mismatching levels types: got $(eltype(levels)), expected " *
+                "$(eltype(c_levels)) based on contrasts levels.",
+            ),
+        )
+    end
+    mismatched_levels = symdiff(c_levels, levels)
+    if !isempty(mismatched_levels)
+        throw(
+            ArgumentError(
+                "contrasts levels not found in data or vice-versa: " *
+                "$mismatched_levels." *
+                "\n  Data levels: $levels." *
+                "\n  Contrast levels: $c_levels",
+            ),
+        )
+    end
+
+    n = length(c_levels)
+    if n == 0
+        throw(
+            ArgumentError(
+                "empty set of levels found (need at least two to compute " * "contrasts)."
+            ),
+        )
```

Comment:
```suggestion
        msg = "empty set of levels found (need at least two to compute " * "contrasts)."
        throw(ArgumentError(msg))
```
The nesting makes the code harder to read IMO